plugins/filter_watermark/watermark.c

```diff
@@ -0,0 +1,392 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2022 The Fluent Bit Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_filter.h>
+#include <fluent-bit/flb_filter_plugin.h>
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_str.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_log.h>
+#include <fluent-bit/flb_time.h>
+#include <msgpack.h>
+#include "stdlib.h"
+
+#include "watermark.h"
+#include "heap.h"
+
+#include <stdio.h>
+#include <sys/types.h>
+
+
+static int configure(struct flb_filter_watermark_ctx *ctx, struct flb_filter_instance *f_ins)
+{
+    const char *str = NULL;
+    double val = 0;
+    char *endp;
+
+    ctx->win_right_edge = 0;;
+    ctx->win_left_edge = 0;
+    ctx->init_flag = 0; 
+
+    /* window size */
+    str = flb_filter_get_property("window_size", f_ins);
+
+    if (str != NULL && (val = strtod(str, &endp)) > 1) {
+        ctx->window_size = val;
+    } 
+    else {
+        ctx->window_size = WINDOW_SIZE_DEFAULT_VALUE;
+    }
+    
+    /* watermark */
+    str = flb_filter_get_property("watermark", f_ins);
+
+    if (str != NULL && (val = strtod(str, &endp)) > 1) {
+        ctx->watermark = val;
+    } 
+    else {
+        ctx->watermark = WATERMARK_DEFAULT_VALUE;
+    }
+
+    /*time_field*/ 
+    str = flb_filter_get_property("time_field", f_ins);
+    
+    if (!str) {
+        flb_plg_error(ctx->ins, "time_field is not defined!");
+        return -1;
+    }
+    ctx->time_field = flb_sds_create(str);
+    return 0;
+}
+
+int deconstructor_imp(void *v) {
```

Comment:
Looks like indentation is not right here and some other places. You can use some C syntax beautifiers to identify/fix.